### PR TITLE
fixed issue where privkey would be unset if relation creation order i…

### LINKED ISSUE #50: `private_key` may end up missing from peer data
### Bug Description

Related alertmanager to self-signed-certificate and saw the error. 
Haven't figured out yet the code path that can lead to `private_key` not being set.

### To Reproduce

1. Deploy alertmanager (2 units)
2. Upgrade alertmanager from path.
3. Relate to self-signed-certificate charm.

### Environment

```
Model                        Controller  Cloud/Region        Version  SLA          Timestamp
test-prometheus-alerts-32k4  uk8s        microk8s/localhost  3.2.0    unsupported  19:10:57-04:00

App   Version  Status   Scale  Charm                     Channel  Rev  Address         Exposed  Message
am    0.25.0   waiting      2  alertmanager-k8s          edge       0  10.152.183.174  no       waiting for units to settle down
ca             active       1  self-signed-certificates  edge       8  10.152.183.38   no       

Unit     Workload  Agent  Address       Ports  Message
am/0     error     idle   10.1.166.78          hook failed: "certificates-relation-joined" for ca:certificates
am/1*    error     idle   10.1.166.80          hook failed: "certificates-relation-joined" for ca:certificates
ca/0*    active    idle   10.1.166.99          
```

### Relevant log output

```shell
unit-am-0: 19:06:31.353 DEBUG unit.am/0.juju-log certificates:7: Emitting Juju event certificates_relation_joined.
unit-am-0: 19:06:31.371 ERROR unit.am/0.juju-log certificates:7: Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 540, in <module>
    main(AlertmanagerCharm, use_juju_for_storage=True)
  File "/var/lib/juju/agents/unit-am-0/charm/venv/ops/main.py", line 441, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-am-0/charm/venv/ops/main.py", line 149, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-am-0/charm/venv/ops/framework.py", line 344, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-am-0/charm/venv/ops/framework.py", line 833, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-am-0/charm/venv/ops/framework.py", line 922, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-am-0/charm/lib/charms/observability_libs/v0/cert_handler.py", line 182, in _on_certificates_relation_joined
    self._generate_csr()
  File "/var/lib/juju/agents/unit-am-0/charm/lib/charms/observability_libs/v0/cert_handler.py", line 213, in _generate_csr
    assert private_key is not None  # for type checker
AssertionError
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by PietroPasotti**
Fixes #50 

I suspect this happens because we don't setup the privkey if, unexpectedly, certificates-created fires before peer-relation-joined does.

### CODE REVIEW

**Review by dstathis — APPROVED** [review]

File: lib/charms/observability_libs/v0/cert_handler.py [hunk 1]
```diff
@@ -224,7 +233,10 @@ def _generate_csr(
         # In case we already have a csr, do not overwrite it by default.
         if overwrite or renew or not self._csr:
             private_key = self._private_key
-            assert private_key is not None  # for type checker
+            if private_key is None:
+                raise RuntimeError(
+                    "private key unset. call _generate_privkey() before you call this method."
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/observability_libs/v0/cert_handler.py:238]: I don't see how to do that without making more changes than I'd like to for this bugfix. That's a larger refactoring

[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/observability_libs/v0/cert_handler.py:238]: that check is a sanity one either way, don't think the current code paths can lead to it

didn't want to make it an assertion though

File: lib/charms/observability_libs/v0/cert_handler.py [hunk 2]
```diff
@@ -168,31 +168,40 @@ def _peer_relation(self) -> Optional[Relation]:
...
+        # todo: compound status would be handy here to set waiting in either case
+        # check peer relation is there
         if not self._peer_relation:
             # tls-certificates relation event happened to fire before peer events.
             # Abort, and let the "peer joined" relation create the CSR.
+            logger.info("certhandler waiting on peer relation")
+            return
+
+        # check cert relation is ready
+        if not (self.charm.model.get_relation(self.certificates_relation_name)):
```
[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/observability_libs/v0/cert_handler.py:198]: not sure what you mean. This feels already very cautious

[the remark this answers is not in the document]
Comment (PietroPasotti) [reply lib/charms/observability_libs/v0/cert_handler.py:198]: and it's out of scope for this pr either way, if you want to open a separate bug for it..

**Review by dstathis — APPROVED** [review]

**Pull request merged**